core/trino-main/src/main/java/io/trino/operator/output/PositionsAppender.java

```diff
@@ -11,37 +11,38 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package io.trino.operator.output;
 
 import io.trino.spi.block.Block;
-import io.trino.spi.block.BlockBuilder;
-import io.trino.spi.type.Type;
+import io.trino.spi.block.BlockBuilderStatus;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 
-import static java.util.Objects.requireNonNull;
+import javax.annotation.Nullable;
 
 public interface PositionsAppender
 {
-    void appendTo(IntArrayList positions, Block source, BlockBuilder target);
-
-    class TypedPositionsAppender
-            implements PositionsAppender
-    {
-        private final Type type;
-
-        public TypedPositionsAppender(Type type)
-        {
-            this.type = requireNonNull(type, "type is null");
-        }
-
-        @Override
-        public void appendTo(IntArrayList positions, Block source, BlockBuilder target)
-        {
-            int[] positionArray = positions.elements();
-            for (int i = 0; i < positions.size(); i++) {
-                type.appendTo(source, positionArray[i], target);
-            }
-        }
-    }
+    void appendTo(IntArrayList positions, Block source);
+
+    Block build();
+
+    /**
+     * Must be called before any {@link #appendRow(Block, int) calls}.
+     * Allows the appender to switch from batch oriented to single row oriented mode if necessary.
+     */
+    void prepareProcessingByRow();
+
+    void appendRow(Block source, int position);
+
+    long getRetainedSizeInBytes();
+
+    /**
+     * Creates new empty {@link PositionsAppender} of the same type as {@code this}.
+     */
+    PositionsAppender newState(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedPositions);
```

Comment:
Rather than creating a new appender and exposing `BlockBuilderStatus`, there should be a `getBlockSizeInBytes` method for computing accumulated block size. `BlockBuilderStatus` can be an internal detail of appenders that actually use `BlockBuilder`